asf_heat_pump_suitability/pipeline/prepare_features/grid_capacity.py

```diff
@@ -0,0 +1,758 @@
+import re
+import os
+import shutil
+from typing import Union, Any
+
+import numpy as np
+import pandas as pd
+import geopandas as gpd
+from shapely.geometry.base import BaseGeometry
+from shapely import wkb
+
+from asf_heat_pump_suitability.getters.s3_getters import (
+    load_s3_data,
+    get_shapefile_from_s3,
+)
+from asf_heat_pump_suitability.getters import get_datasets
+
+# Constants
+CRS = "EPSG:4326"  # Geometry coordinate reference system - standard longitude/latitude projection
+TAKEUP_RATE = 30  # Percentage of households in each LSOA to install heat pumps
+POWER_PER_HEATPUMP = 8  # Power rating per heat pump in kW
+COEFFICIENT_OF_PERFORMANCE = (
+    2.5  # Heat output (heat pump rating) / electricity consumed
+)
+SUBSTATION_SCALING_FACTOR = 1.0  # Factor to scale substation power rating (MVA) by
+
+
+def _parse_binary_geometry(
+    binary_data: Union[BaseGeometry, bytes, str]
+) -> Union[BaseGeometry, None]:
+    """
+    Parse binary geometry data into Shapely geometry object.
+
+    Args:
+        binary_data: Input geometry data in various formats.
+
+    Returns:
+        Shapely geometry object or None if parsing fails.
+    """
+    if isinstance(binary_data, BaseGeometry):
+        return binary_data
+    elif isinstance(binary_data, bytes):
+        return wkb.loads(binary_data)
+    elif isinstance(binary_data, str):
+        return wkb.loads(binary_data, hex=True)
+    else:
+        return None
+
+
+def _parse_capacity(value: Any) -> float:
+    """
+    Parse complex capacity strings and return total capacity as float.
+
+    Args:
+        value: Input capacity value in various formats.
+
+    Returns:
+        Parsed capacity as a float, or NaN if parsing fails.
+    """
+    if pd.isna(value):
+        return np.nan
+
+    if isinstance(value, (int, float)):
+        return float(value)
+
+    value = str(value).strip().lower()
+
+    # Pattern to match "X x Y" format (e.g., "2 x 12.5")
+    pattern = r"(\d+(?:\.\d+)?)\s*x\s*(\d+(?:\.\d+)?)"
+    match = re.match(pattern, value)
+
+    if match:
+        count = float(match.group(1))
+        individual_capacity = float(match.group(2))
+        return count * individual_capacity
+
+    try:
+        return float(value)
+    except ValueError:
+        print(f"Warning: Could not parse capacity value: {value}")
```

Comment:
```suggestion
        logging.warning(f"Could not parse capacity value: {value}")
```
need to import logging at top of script as well